File: UEReflectWatch/Projectscanner.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace UEReflectWatch
{
    public sealed class ProjectScanResult
    {
        public int FilesScanned { get; set; }
        public int FilesWithMacros { get; set; }
        public int TotalMacros { get; set; }
        public List<string> Errors { get; } = new List<string>();
    }

    public static class ProjectScanner
    {
        // Folders that are never part of the user's own source. Scanning them
        // would seed thousands of engine header entries into the state store
        // which wastes memory, slows every subsequent diff, and would generate
        // false positives if the engine itself is ever updated.
        private static readonly HashSet<string> ExcludedFolderNames = new HashSet<string>(
            StringComparer.OrdinalIgnoreCase)
        {
            "Binaries",
            "Intermediate",
            "DerivedDataCache",
            "Saved",
            "Plugins",   // exclude plugins to keep scope tight; can be made optional later
            ".vs",
            ".git"
        };

        public static ProjectScanResult ScanAndSeed(string projectRootDir, StateStore stateStore)
        {
            var result = new ProjectScanResult();

            var sourceDir = Path.Combine(projectRootDir, "Source");
            if (!Directory.Exists(sourceDir))
            {
                // Fall back to the project root if there is no Source subfolder.
                sourceDir = projectRootDir;
            }

            var headerFiles = FindHeaderFiles(sourceDir, result);

            foreach (var filePath in headerFiles)
            {
                try
                {
                    var content = File.ReadAllText(filePath);
                    var macros = MacroScanner.Scan(content);
                    stateStore.SetMacros(filePath, macros);

                    result.FilesScanned++;
                    if (macros.Count > 0)
                    {
                        result.FilesWithMacros++;
                        result.TotalMacros += macros.Count;
                    }
                }
                catch (Exception ex)
                {
                    result.Errors.Add($"{Path.GetFileName(filePath)}: {ex.Message}");
                }
            }

            return result;
        }

        private static List<string> FindHeaderFiles(string rootDir, ProjectScanResult result)
        {
            var files = new List<string>();

            try
            {
                // Non-recursive enumeration so we can filter excluded folders.
                foreach (var file in Directory.GetFiles(rootDir, "*.h", SearchOption.TopDirectoryOnly))
                {
                    files.Add(file);
                }

                foreach (var subDir in Directory.GetDirectories(rootDir))
                {
                    var folderName = Path.GetFileName(subDir);
                    if (ExcludedFolderNames.Contains(folderName)) continue;

                    files.AddRange(FindHeaderFiles(subDir, result));
                }
            }
            catch (Exception ex)
            {
                result.Errors.Add($"Could not read directory {rootDir}: {ex.Message}");
            }

            return files;
        }
    }
}

File: RiderPlugin/src/main/kotlin/com/ueReflectWatch/resolver/ProjectScanner.kt
package com.ueReflectWatch.resolver

import com.ueReflectWatch.scanner.MacroScanner
import com.ueReflectWatch.store.StateStore
import java.io.File

data class ProjectScanResult(
    val filesScanned: Int,
    val filesWithMacros: Int,
    val totalMacros: Int,
    val errors: List<String>
)

object ProjectScanner {

    private val EXCLUDED_FOLDERS = setOf(
        "Binaries", "Intermediate", "DerivedDataCache",
        "Saved", "Plugins", ".vs", ".git"
    )

    fun scanAndSeed(projectRootDir: String, stateStore: StateStore): ProjectScanResult {
        var filesScanned = 0
        var filesWithMacros = 0
        var totalMacros = 0
        val errors = mutableListOf<String>()

        val sourceDir = File(projectRootDir, "Source")
            .takeIf { it.exists() } ?: File(projectRootDir)

        val headerFiles = findHeaderFiles(sourceDir, errors)

        for (file in headerFiles) {
            try {
                val content = file.readText()
                val macros = MacroScanner.scan(content)
                stateStore.setMacros(file.absolutePath, macros)
                filesScanned++
                if (macros.isNotEmpty()) {
                    filesWithMacros++
                    totalMacros += macros.size
                }
            } catch (e: Exception) {
                errors.add("${file.name}: ${e.message}")
            }
        }

        return ProjectScanResult(filesScanned, filesWithMacros, totalMacros, errors)
    }

    private fun findHeaderFiles(dir: File, errors: MutableList<String>): List<File> {
        val results = mutableListOf<File>()

        try {
            dir.listFiles()?.forEach { entry ->
                when {
                    entry.isFile && entry.extension == "h" -> results.add(entry)
                    entry.isDirectory && entry.name !in EXCLUDED_FOLDERS ->
                        results.addAll(findHeaderFiles(entry, errors))
                }
            }
        } catch (e: Exception) {
            errors.add("Could not read directory ${dir.path}: ${e.message}")
        }

        return results
    }
}
